Conan-elisp Reciepe

Carl2 opened the pull request:
Recipe for conan-elisp.

### Brief summary of what the package does

The purpose of this elisp-package is to enable the use of _conan 2.0_ libraries to retain flags for c++ in elisp. This is particularly good when using for example babel in org-mode to make small snippets of c++ code that uses some external library.



### Direct link to the package repository

https://github.com/Carl2/conan-elisp


### Your association with the package

maintainer, contributor



### Checklist


- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] I've used `M-x checkdoc` to check the package's documentation strings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

jcs090218 commented:
I'm helping MELPA maintainers a little bit here.

`conan-elisp.el` with byte-compile using Emacs 29.1:

```
conan-elisp.el:79:2: Warning: Package cl is deprecated
```

`conan-elisp.el` with checkdoc (0.6.2)

```
No issues found
```

`conan-elisp.el` with package-lint (20230905.629)

```
1 issue found:

79:29: warning: Replace deprecated `cl' with `cl-lib'.  The `cl-libify' package can help with this.
```

CalleOlsen commented:
Thank you!
I resolved the issue with the warning! Sorry about that, should be fixed now.

riscy commented:
Thank you both - a couple small notes in addition:

`conan-elisp.el` with package-lint 20230905.629:
```
1 issue found:
79:10: warning: This file is not in the `cl-lib' ELPA compatibility package: require `cl-lib' instead.
```
You should be able to just `(require 'cl-lib)` and get this one out of the way.

`conan-elisp.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- conan-elisp.el#L26: Triple semicolons `;;;` are usually for section headings
```
Minor, but you should drop the triple semicolons for these paragraphs.


I also notice you have a few examples at the bottom of the file:
```elisp
;; Examples
;;(conan-elisp-install "fmt/10.1.1 sml/1.1.6 zlib/1.2.13" 'all)
;; With pre flags
;;(conan-elisp-install "fmt/10.1.1 sml/1.1.9" 'all "-std=c++20 -Wall -Wextra")
;; With post flags
;; (conan-elisp-install "fmt/10.1.1 sml/1.1.9" 'all nil "-std=c++20 -Wall -Wextra")
;; With both pre and post flags
;; (conan-elisp-install "fmt/10.1.1 sml/1.1.9" 'all "-std=c++20 -Wall -Wextra" "-o /tmp/myElf")
```
maybe these would make sense in your `;;; Commentary` instead?

CalleOlsen commented:
Thank you! 
Cool , didn't know about `;;;` being sections, and [melpazoid](https://github.com/riscy/melpazoid).
Sorry for the issues, i'm on a bit of a learning curve (both with melpa and emacs-lisp).
I fixed the above issues and ran _melpazoid_  locally:

```
python3 melpazoid/melpazoid.py
Enter recipe or URL for MELPA PR: https://github.com/melpa/melpa/pull/8756
<!-- Checking pull request: https://github.com/melpa/melpa/pull/8756 -->
<!-- Cloning https://github.com/carl2/conan-elisp.git  -->
<!-- Building container for conan-elisp... 🐳 -->
make[1]: Entering directory '/home/calle/git/melpazoid'
sha256:4523b2c714292f6f5aaf47529ba0c5262e302a32a94e017578cd4a37c7e84045

Package and license:
- MIT License (via API)
- GPL* -- conan-elisp.el: Generate header for c++


python3 melpazoid/melpazoid.py
Enter recipe or URL for MELPA PR: https://github.com/melpa/melpa/pull/8756
<!-- Checking pull request: https://github.com/melpa/melpa/pull/8756 -->
<!-- Cloning https://github.com/carl2/conan-elisp.git  -->
<!-- Building container for conan-elisp... 🐳 -->
make[1]: Entering directory '/home/calle/git/melpazoid'
sha256:4523b2c714292f6f5aaf47529ba0c5262e302a32a94e017578cd4a37c7e84045

Package and license:
- MIT License (via API)
- GPL* -- conan-elisp.el: Generate header for c++

<!-- Footnotes:
- (conan-elisp
    :fetcher github
    :repo "carl2/conan-elisp")
- Created: 2023-03-25T10:18:10Z
- Updated: 2023-06-16T20:05:51Z
- Watched: 2
-->

<!-- Finished -->
```

tarsius commented:
>    Package and license:
>    - MIT License (via API)
>    - GPL* -- conan-elisp.el: Generate header for c++

The `LICENSE` file says that the license is `MIT`, but the permission statement in the library says that it is `GPL-3.0-or-later`. Please decide which one you want to use, and correct the information in the other place. If you would like to use multiple licenses, then use an appropriate `SPDX-License-Identifier` header as explained at https://spdx.dev/learn/handling-license-info/#how.

Carl2 commented:
Sorry, 
I changed the permission statement  to `MIT` too should be ok now.

riscy commented:
Good catch on that license - I missed that one.  I think the checklist look great now.

Let's talk about the package name.  I realize now that it's probably called `conan-elisp` because this is the "Conan" utility made accessible through an elisp implementation.  But I think it makes much more sense just to call this package `conan`.   You could leave the repo named "conan-elisp" but I think the package itself (the elisp files, the prefixes, etc.) should all just be `conan`.

Carl2 commented:
Certainly,
I agree, `conan` would probably be a much better name. I will fix that asap.

Carl2 commented:
Renamed _conan-elisp_ to _conan_ instead.

- [x] Updated the code to use just `conan-install`
- [x] Update documentation
- [x] Update Melpa PR
- [x] Check lint, checkdoc and byte-compile
- [x] Ran through with _melpazoid_

```
Package and license:
- MIT License (via API)
- MIT License -- conan.el: Generate flags for  for c++ using conan 2.0

<!-- Footnotes:
- (conan
    :fetcher github
    :repo "carl2/conan-elisp")
- Created: 2023-03-25T10:18:10Z
- Updated: 2023-10-02T09:24:31Z
- Watched: 2
-->

<!-- Finished -->
```

riscy commented:
Thanks again.  Sorry for creating a delay on this, but I notice a couple of other small issues as I take another look:

Most importantly you're using `shell-command-to-string` in a couple of places, where you're feeding commands in unescaped.  So for example in `conan-conan-get-compile-flags` if `output-path` contains shell control characters like `|`, `&`, etc (or even just a space) then you'll get unexpected and sometimes dangerous results.  Use `shell-quote-argument` to fix this.

Incidentally I think this same function is running `(message "cmd %s " (concat cmd lib-no-ver))` which may be a leftover print-debug statement - just fyi.

I also notice in `conan-install`'s docstring you have:
```
- `flags' could be either \='include, \='libs, \='all, or \='both
```
but you actually want two slashes to properly escape these:
```
- `flags' could be either \\='include, \\='libs, \\='all, or \\='both
```

Carl2 commented:
Great remarks and no worries!
I will fix that as soon as possible.

Cheers!

Carl2 commented:
- [x] Updated code to include checks on argument for `shell-command-to-string` using `shell-quote-arguments` for output paths.
- [x] Update documentation
- [x] Check lint, checkdoc and byte-compile
- [x] Ran through with _melpazoid_

Cheers

riscy commented:
Thank you again for your patience!  Merging this now, and welcome to MELPA. :)
